Customisable navigation bar (back-end)

sergei-kucher opened the pull request:
## Checklist

- [ ] Added entry to changelog
<details>
<summary>More info</summary>
Add a concise line to the 'Next release' section of the changelog (docs/README.md) so people other than developers can understand what has changed where. E.g. 'Added an error message to the project name field of the project edit form (Admin > Projects > Edit)'.
</details>

- [ ] Tests
<details>
<summary>More info</summary>

### Unit tests

Did you add relevant unit tests?

### E2E tests

Sometimes it can be more efficient to update E2E tests after CI has run them. If you know which ones to update, go ahead! E2E template cl2-back:

```bash
docker compose run --rm web bin/rails cl2_back:create_tenant[localhost,e2etests_template]
```

</details>

- [ ] Prepared branch for code review
<details>
<summary>More info</summary>
Reviewed code to reduce unnecessary back and forth (removal of console.log, comments, ...)? Added comments to clarify code, emphasize what to pay attention to, etc.?
</details>

## Links

- [Jira ticket](**put URL here**)
- [citizenlab-ee PR](**put URL here** or remove)
- [Specs](**put URL here** or remove)
- [Epic Deployment](**put URL here** or remove)

## What changes are in this PR?

Add a concise summary here of what happened, so the reviewer has some background.

## How urgent is a code review?

Let the reviewer(s) know how urgent the code review is, so they can prioritize their work accordingly. Be specific (e.g. by Wednesday, end of the day/this week/... is better than 'urgent' or 'very urgent'). Optionally provide a word of explanation on your deadline.

---

- [ ] Change back the postgres image in docker-compose
- [ ] Don't merge to master. Create a branch for customizable navbar

sergei-kucher committed:
Customisable navigation bar

luucvanderzee committed:
Split 'LEGAL_PAGES' into 'FIXED_PAGES' and 'POLICY_PAGES'

luucvanderzee committed:
Added 'policies' tab

luucvanderzee committed:
Correctly use formatting in FormattedMessage

luucvanderzee committed:
Pages/Policies subtitle link underline

Commit:
Translations updated by CI (extract-intl)

luucvanderzee committed:
Updated CHANGELOG

luucvanderzee committed:
Merge branch 'CL2-6676-i3-create-policies-tab' of https://github.com/CitizenLabDotCo/citizenlab into CL2-6676-i3-create-policies-tab

luucvanderzee committed:
Improved CHANGELOG description

luucvanderzee committed:
Merge branch 'master' into CL2-6676-i3-create-policies-tab

luucvanderzee committed:
_

luucvanderzee commented:
Hey, I noticed that the the order of the nav bar items is determined by the `position` column. I was wondering if it makes sense to change this to `ordering`? Because that is the terminology used in other parts of the code base. It would also save me some work on the frontend, since the drag and drop components also expect an `ordering` attribute (e.g. `page.attributes.ordering`).

sergei-kucher committed:
done

adessy committed:
IN-552 Add color_index property to nodes

IN-552 #time 0w 0d 4h 46m Add color_index property to nodes

brentguf committed:
Up

brentguf committed:
Remove line

brentguf committed:
Next release

brentguf committed:
Change release data

kogre committed:
New translations en.json (Serbian (Latin))

sergei-kucher committed:
specs

adessy committed:
IN-557 Fix "stack level too deep" errors generated when capturing exceptions with Sentry

IN-557 #time 0w 2d 5h 46m Fix "stack level too deep" errors generated when capturing exceptions with Sentry

adessy committed:
IN-552 Remove dangling debugger instruction

IN-552 #time 0w 0d 0h 23m Remove dangling debugger instruction

adessy committed:
IN-552 Add tests for node colors

IN-552 #time 0w 0d 0h 35m Add tests for node colors

adessy committed:
IN-552 Fix broken tests

IN-552 #time 0w 0d 0h 15m Fix broken tests

luucvanderzee committed:
Timeline: behavior wrt very short phases

luucvanderzee committed:
Updated CHANGELOG

luucvanderzee committed:
Improved naming of container padding constant

luucvanderzee committed:
Made CSS more logical

adessy committed:
Merge pull request #751 from CitizenLabDotCo/IN-557/sentry-stack-too-deep

IN-557 Fix "stack level too deep" errors when capturing exceptions with Sentry

kogre committed:
Don't load rack-mini-profiler in production

adessy committed:
Merge pull request #753 from CitizenLabDotCo/dont-load-rank-mini-profiler-in-prod

Don't load rack-mini-profiler in production

jinjagit committed:
Add locale to CL2_SUPPORTED_LOCALES

jinjagit committed:
Add locale to constants.js & .ts

jinjagit committed:
Create lb-LU.ts

IvaKop committed:
Update library version

brentguf committed:
Merge branch 'master' of https://github.com/CitizenLabDotCo/citizenlab into changelog/release/2021-09-17

brentguf committed:
Change release date

brentguf committed:
Merge pull request #738 from CitizenLabDotCo/changelog/release/2021-09-17

Changelog/release/2021 09 20

adessy committed:
Fix header in CHANGELOG.md

brentguf committed:
Merge pull request #749 from CitizenLabDotCo/crowdin_master

New Crowdin updates

luucvanderzee committed:
Merge pull request #752 from CitizenLabDotCo/CL2-6611-equidistant-timeline-phases

[CL2-6611] Equidistant timeline phases

jinjagit committed:
Merge pull request #755 from CitizenLabDotCo/CL2-6739-Enable-Luxembourgish

CL2-6739 Add Luxembourgish (lb-LU) as a platform language

kogre committed:
New translations en.json (French)

kogre committed:
New translations en.json (French, Belgium)

kogre committed:
New translations en.json (Dutch, Belgium)

kogre committed:
New translations en.json (English, United Kingdom)

kogre committed:
New translations en.json (Dutch)

kogre committed:
New translations en.json (Spanish)

kogre committed:
New translations en.json (German)

kogre committed:
New translations en.json (English, Canada)

kogre committed:
New translations en.json (English)

kogre committed:
New translations en.json (Spanish, Chile)

kogre committed:
Update source file en.json

kogre committed:
New translations en.json (English)

kogre committed:
Update source file en.json

kogre committed:
New translations en.json (Portuguese, Brazilian)

luucvanderzee committed:
Reverted upgrade js-cookie

kogre committed:
New translations en.json (Serbian (Latin))

luucvanderzee committed:
Merge pull request #761 from CitizenLabDotCo/revert-upgrade-js-cookie

Reverted upgrade js-cookie

kogre committed:
New translations en.json (Serbian (Latin))

IvaKop committed:
Merge branch 'master' into IN-559/viz-colors

IvaKop committed:
Add colors

IvaKop committed:
Move function out of jsx

IvaKop committed:
Disable eslint rule for T - to be fixed later

adessy committed:
Merge branch 'master' into IN-552/networks-with-color-codes

IN-552 #time 0w 1d 3h 57m Merge branch 'master' into IN-552/networks-with-color-codes

kogre committed:
New translations en.json (Portuguese, Brazilian)

kogre committed:
New translations en.yml (Portuguese, Brazilian)

kogre committed:
New translations en.yml (Portuguese, Brazilian)

IvaKop committed:
Add SmartSurvey integration

IvaKop committed:
Update changelog

Commit:
Translations updated by CI (extract-intl)

kogre committed:
New translations en.json (Portuguese, Brazilian)

kogre committed:
New translations en.yml (Portuguese, Brazilian)

IvaKop committed:
Fix lint error

IvaKop committed:
Merge branch 'IN-556/smartsurvey-integration' of github.com:CitizenLabDotCo/citizenlab into IN-556/smartsurvey-integration

luucvanderzee committed:
Updated CHANGELOG

luucvanderzee committed:
Updated CHANGELOG (again)

luucvanderzee committed:
Fix linter no-children-prop error

kogre commented:
Hey @sergey-kucher , what is the easiest for you, in terms of review?
Should I still do it on this PR, or in another branch where FE is merged?

kogre committed:
New translations en.json (French)

kogre committed:
New translations en.json (German)

kogre committed:
New translations en.json (Dutch)

kogre committed:
New translations en.json (Norwegian)

kogre committed:
New translations en.json (Spanish)

kogre committed:
New translations en.json (Portuguese, Brazilian)

IvaKop committed:
Improve nodeColor method

IvaKop committed:
Delete mistakenly commited node modules folder

kogre committed:
New translations en.json (Danish)

kogre commented on back/app/models/navbar_item.rb:
```diff
@@ -0,0 +1,106 @@
+    end
+  end
+
+  def ordering_cannot_be_aside_from_others
+    if ordering > (self.class.where(visible: visible).maximum_ordering + 1)
+      errors.add(:ordering, "cannot be aside from others")
+    end
+  end
```
Have bad experience with such validations (that are not just about the current record, but depend on other records). These will work as desired in most scenarios, but will break when e.g. you create some items in bulk, delete an item in-between, etc.

In the rest of the code, we're using `acts_as_list` gem for it. It works pretty well, and by itself deals with list-related validations. Not sure whether you saw that, but if you did, any reason not to adopt it here?

kogre commented on back/app/models/navbar_item.rb:
```diff
@@ -0,0 +1,106 @@
+
+  def cannot_destroy_reserved
+    if ordering <= LAST_RESERVED_ORDERING
+      errors.add :ordering,
+        "Cannot destroy a reserved navbar item. Ordering (#{ordering}) should be > #{LAST_RESERVED_ORDERING}."
```
Since our error in the end reach the FE, we work with codes instead of text. Or both, but at least we need the code, so the FE can interpret it and show the right (multilingual) copy.

So this should become something like `errors.add(:ordering, :cannot_destroy_reserved, message: "Cannot destr... ")`

kogre commented on back/app/models/navbar_item.rb:
```diff
@@ -0,0 +1,106 @@
+    end
+
+    unless FEATURES.fetch(:destroy).include?(type)
+      errors.add :type,
+        "Cannot destroy a reserved navbar item. Navbar items with type #{type} impossible to remove."
```
Same here.

kogre commented on back/app/models/navbar_item.rb:
```diff
@@ -0,0 +1,106 @@
+    return unless ordering_changed?
+
+    if type_was && !FEATURES.fetch(:reorder).include?(type_was)
+      errors.add :type, "Cannot reorder a reserved item. It's not allowed for the type (#{type_was})"
```
And here... Won't repeat further, you get the picture :)

kogre commented on back/app/models/navbar_item.rb:
```diff
@@ -0,0 +1,106 @@
+    errors.add :title_multiloc, "Cannot edit title for type '#{type_was}'" if title_multiloc_changed?
+  end
+
+  def list_of_visible_items_is_already_full
+    visible_items_count = NavbarItem.where(visible: true).count
+    return if visible_items_count < MAX_VISIBLE_ITEMS
+
+    errors.add :visible, "Cannot make the item visible when the list of visible items is full (max: #{MAX_VISIBLE_ITEMS})"
+  end
```
Same here, validations depending on other records have a tendency to be problematic. They are prone to race conditions like bulk updates, or cached queries. If you're confident it works in all cases performed by the API, I can live with it.

kogre commented on back/app/models/page.rb:
```diff
@@ -28,6 +35,55 @@ def published?
+    return unless slug_changed?
+
+    if navbar_item.nil?
+      errors.add :slug, <<~TXT.split("\n").join(" ")
```
Using codes for errors in this file as well.

kogre reviewed (commented):
Hey @sergey-kucher , didn't get through it all yet, but already submitting what I have.

kogre committed:
New translations en.json (Italian)

adessy committed:
Merge pull request #750 from CitizenLabDotCo/IN-552/networks-with-color-codes

IvaKop committed:
Merge branch 'master' into IN-559/viz-colors

IvaKop committed:
Merge pull request #762 from CitizenLabDotCo/IN-559/viz-colors

IN-559 Add colors to Insights visualization

IvaKop committed:
Merge pull request #756 from CitizenLabDotCo/bump-component-library-version

[NO-TICKET] Bump component library version

IvaKop committed:
Merge master

IvaKop committed:
Merge pull request #763 from CitizenLabDotCo/IN-556/smartsurvey-integration

IN-556 Add SmartSurvey integration

kogre committed:
New translations en.json (Romanian)

kogre committed:
New translations en.json (English)

kogre committed:
New translations en.json (French, Belgium)

kogre committed:
New translations en.json (Dutch, Belgium)

kogre committed:
New translations en.json (Serbian (Latin))

kogre committed:
New translations en.json (Luxembourgish)

kogre committed:
New translations en.json (English, United Kingdom)

kogre committed:
New translations en.json (English, Canada)

kogre committed:
New translations en.json (Maori)

kogre committed:
New translations en.json (Spanish, Chile)

kogre committed:
New translations en.json (Portuguese, Brazilian)

kogre committed:
New translations en.json (Swedish)

kogre committed:
New translations en.json (French)

kogre committed:
New translations en.json (Polish)

kogre committed:
New translations en.json (Norwegian)

kogre committed:
New translations en.json (Dutch)

kogre committed:
New translations en.json (Italian)

kogre committed:
New translations en.json (Hungarian)

kogre committed:
New translations en.json (German)

kogre committed:
New translations en.json (Danish)

kogre committed:
New translations en.json (Arabic)

kogre committed:
New translations en.json (Spanish)

kogre committed:
New translations en.json (Greenlandic)

IvaKop committed:
Merge pull request #758 from CitizenLabDotCo/crowdin_master

New Crowdin updates

kogre committed:
Update source file en.json

kogre committed:
New translations en.json (Romanian)

kogre committed:
New translations en.json (English)

kogre committed:
New translations en.json (French, Belgium)

kogre committed:
New translations en.json (Serbian (Latin))

kogre committed:
New translations en.json (Luxembourgish)

kogre committed:
New translations en.json (Maori)

kogre committed:
New translations en.json (Spanish, Chile)

kogre committed:
New translations en.json (Portuguese, Brazilian)

kogre committed:
New translations en.json (Swedish)

kogre committed:
New translations en.json (French)

kogre committed:
New translations en.json (Polish)

kogre committed:
New translations en.json (Norwegian)

kogre committed:
New translations en.json (Italian)

kogre committed:
New translations en.json (Hungarian)

kogre committed:
New translations en.json (German)

kogre committed:
New translations en.json (Danish)

kogre committed:
New translations en.json (Arabic)

kogre committed:
New translations en.json (Spanish)

kogre committed:
New translations en.json (Greenlandic)

IvaKop committed:
Merge pull request #771 from CitizenLabDotCo/crowdin_master

New Crowdin updates

IvaKop committed:
Update changelog

IvaKop committed:
Merge pull request #773 from CitizenLabDotCo/update-changelog-2021-09-24

Update changelog

IvaKop committed:
Merge pull request #766 from CitizenLabDotCo/delete-and-ignore-node-modules

[NO-TICKET] Delete mistakenly committed node modules folder

kogre committed:
New translations en.json (Portuguese, Brazilian)

kogre committed:
New translations en.yml (Portuguese, Brazilian)

kogre committed:
New translations en.json (Spanish)

dependabot[bot] committed:
Bump @types/jwt-decode from 2.2.1 to 3.1.0 in /front

Bumps [@types/jwt-decode](https://github.com/auth0/jwt-decode) from 2.2.1 to 3.1.0.
- [Release notes](https://github.com/auth0/jwt-decode/releases)
- [Changelog](https://github.com/auth0/jwt-decode/blob/master/CHANGELOG.md)
- [Commits](https://github.com/auth0/jwt-decode/commits/v3.1.0)

---
updated-dependencies:
- dependency-name: "@types/jwt-decode"
  dependency-type: direct:development
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump react-medium-image-zoom from 3.1.3 to 4.3.5 in /front

Bumps [react-medium-image-zoom](https://github.com/rpearce/image-zoom) from 3.1.3 to 4.3.5.
- [Release notes](https://github.com/rpearce/image-zoom/releases)
- [Changelog](https://github.com/rpearce/image-zoom/blob/main/CHANGELOG.md)
- [Commits](https://github.com/rpearce/image-zoom/commits/v4.3.5)

---
updated-dependencies:
- dependency-name: react-medium-image-zoom
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump babel-plugin-styled-components from 1.12.0 to 1.13.2 in /front

Bumps [babel-plugin-styled-components](https://github.com/styled-components/babel-plugin-styled-components) from 1.12.0 to 1.13.2.
- [Release notes](https://github.com/styled-components/babel-plugin-styled-components/releases)
- [Commits](https://github.com/styled-components/babel-plugin-styled-components/compare/v1.12.0...v1.13.2)

---
updated-dependencies:
- dependency-name: babel-plugin-styled-components
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump glob from 7.1.7 to 7.2.0 in /front

Bumps [glob](https://github.com/isaacs/node-glob) from 7.1.7 to 7.2.0.
- [Release notes](https://github.com/isaacs/node-glob/releases)
- [Changelog](https://github.com/isaacs/node-glob/blob/master/changelog.md)
- [Commits](https://github.com/isaacs/node-glob/compare/v7.1.7...v7.2.0)

---
updated-dependencies:
- dependency-name: glob
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump jest-cli from 26.6.3 to 27.2.2 in /front

Bumps [jest-cli](https://github.com/facebook/jest/tree/HEAD/packages/jest-cli) from 26.6.3 to 27.2.2.
- [Release notes](https://github.com/facebook/jest/releases)
- [Changelog](https://github.com/facebook/jest/blob/main/CHANGELOG.md)
- [Commits](https://github.com/facebook/jest/commits/v27.2.2/packages/jest-cli)

---
updated-dependencies:
- dependency-name: jest-cli
  dependency-type: direct:development
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump jwt-decode from 2.2.0 to 3.1.2 in /front

Bumps [jwt-decode](https://github.com/auth0/jwt-decode) from 2.2.0 to 3.1.2.
- [Release notes](https://github.com/auth0/jwt-decode/releases)
- [Changelog](https://github.com/auth0/jwt-decode/blob/master/CHANGELOG.md)
- [Commits](https://github.com/auth0/jwt-decode/compare/v2.2.0...v3.1.2)

---
updated-dependencies:
- dependency-name: jwt-decode
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump regenerator-runtime from 0.13.7 to 0.13.9 in /front

Bumps [regenerator-runtime](https://github.com/facebook/regenerator) from 0.13.7 to 0.13.9.
- [Release notes](https://github.com/facebook/regenerator/releases)
- [Commits](https://github.com/facebook/regenerator/compare/regenerator-runtime@0.13.7...regenerator-runtime@0.13.9)

---
updated-dependencies:
- dependency-name: regenerator-runtime
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump typescript from 4.3.2 to 4.4.3 in /front

Bumps [typescript](https://github.com/Microsoft/TypeScript) from 4.3.2 to 4.4.3.
- [Release notes](https://github.com/Microsoft/TypeScript/releases)
- [Commits](https://github.com/Microsoft/TypeScript/compare/v4.3.2...v4.4.3)

---
updated-dependencies:
- dependency-name: typescript
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

luucvanderzee committed:
Fix typescript errors

luucvanderzee committed:
Merge pull request #779 from CitizenLabDotCo/dependabot/npm_and_yarn/front/glob-7.2.0

Bump glob from 7.1.7 to 7.2.0 in /front

luucvanderzee committed:
Merge pull request #782 from CitizenLabDotCo/dependabot/npm_and_yarn/front/regenerator-runtime-0.13.9

Bump regenerator-runtime from 0.13.7 to 0.13.9 in /front

luucvanderzee committed:
Merge pull request #778 from CitizenLabDotCo/dependabot/npm_and_yarn/front/babel-plugin-styled-components-1.13.2

Bump babel-plugin-styled-components from 1.12.0 to 1.13.2 in /front

luucvanderzee committed:
Merge pull request #780 from CitizenLabDotCo/dependabot/npm_and_yarn/front/jest-cli-27.2.2

Bump jest-cli from 26.6.3 to 27.2.2 in /front

kogre committed:
New translations en.json (Portuguese, Brazilian)

kogre committed:
New translations en.json (Spanish)

brentguf committed:
Merge pull request #774 from CitizenLabDotCo/crowdin_master

New Crowdin updates

sergei-kucher committed:
-

jinjagit committed:
Remove skip statements, added in PR #566, from tests

luucvanderzee committed:
Merge pull request #783 from CitizenLabDotCo/dependabot/npm_and_yarn/front/typescript-4.4.3

Bump typescript from 4.3.2 to 4.4.3 in /front

luucvanderzee committed:
Merge branch 'master' into CL2-6676-i3-create-policies-tab

Commit:
Translations updated by CI (extract-intl)

luucvanderzee committed:
Merge pull request #719 from CitizenLabDotCo/CL2-6676-i3-create-policies-tab

CL2-6676/i3: Create policies tab

jinjagit committed:
Merge pull request #786 from CitizenLabDotCo/CL2-6685-remove-skip-from-tests-on-this-ticket

CL2-6685 Remove skip statements, added in PR #566, from tests

luucvanderzee committed:
Merge pull request #781 from CitizenLabDotCo/dependabot/npm_and_yarn/front/jwt-decode-3.1.2

Bump jwt-decode from 2.2.0 to 3.1.2 in /front

luucvanderzee committed:
Merge pull request #776 from CitizenLabDotCo/dependabot/npm_and_yarn/front/types/jwt-decode-3.1.0

Bump @types/jwt-decode from 2.2.1 to 3.1.0 in /front

luucvanderzee committed:
Removed react-medium-image-zoom

luucvanderzee committed:
Merge pull request #777 from CitizenLabDotCo/dependabot/npm_and_yarn/front/react-medium-image-zoom-4.3.5

Bump react-medium-image-zoom from 3.1.3 to 4.3.5 in /front

sergei-kucher commented:
@kogre Something went wrong with this PR. The diffs are wrong. Looks like some kind if issue on github.

I had to open new pull-request (https://github.com/CitizenLabDotCo/citizenlab/pull/793/).
I've moved your comments there.
